### PR TITLE
Bigendian fixes: DEX parser

### PR DESCRIPTION
**Pull request by zamaudio**

### CODE REVIEW

**Comment by radare** [comment]
looks like there's a segfault for gcc
```
[33m[  ]  103  fuzzed: [0mOther Fuzz aa ; q ../../bins/fuzzed/fuzz-loaduleb.dex
[31m[XX][0m

[35mRegression: [0mrun_tests.sh
[35mCommand: [0m /home/travis/build/radare/radare2/install/bin/radare2 -e scr.color=0 -N -q -i /tmp/r2-regressions//fuzzed-5gVdvb/rad   ../../bins/fuzzed/fuzz-loaduleb.dex > /tmp/r2-regressions//fuzzed-5gVdvb/out 2> /tmp/r2-regressions//fuzzed-5gVdvb/err
[35mFile: [0m../../bins/fuzzed/fuzz-loaduleb.dex
[35mScript: [0maaa ; q
Segmentation fault (core dumped)

```

**Comment by radare** [comment]
the same code with clang doesnt crashes. can u run it with valgrind or soemthing to fix the regression?

**Comment by zamaudio** [comment]
I cant reproduce the segfault on my machine with gcc build..,

**Comment by radare** [comment]
try with asan or valgrind

**Comment by zamaudio** [comment]
Found a bug with valgrind... hope this passes now

File: libr/bin/format/dex/dex.c [hunk 1]
```diff
@@ -25,14 +26,32 @@ struct r_bin_dex_obj_t* r_bin_dex_new_buf(RBuffer *buf) {
...
 	}
-	// XXX: this is not endian safe
-	// XXX: no need to dup all data!! just pointers to the bin->b
-	// XXX: return value not checked
 	/* header */
-	//r_buf_read_at (bin->b, 0, (ut8*)&bin->header, sizeof (struct dex_header_t));
 	if (bin->size < sizeof(struct dex_header_t))
 		goto fail;
-	bin->header = (*(struct dex_header_t*)bin->b->buf);
+	memcpy (&bin->header.magic, bin->b->buf, 8);
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: this hurts my r_buf_fread_at pride

File: libr/bin/format/dex/dex.c [hunk 2]
```diff
@@ -25,14 +26,32 @@ struct r_bin_dex_obj_t* r_bin_dex_new_buf(RBuffer *buf) {
...
+	bin->header.checksum = r_read_le32 (bin->b->buf + 8);
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: add a new pointer that just points to bin->b->buf. there's no need to dereference two pointers on every single line

File: libr/bin/format/dex/dex.c [hunk 3]
```diff
@@ -56,10 +78,26 @@ struct r_bin_dex_obj_t* r_bin_dex_new_buf(RBuffer *buf) {
 	}
 	bin->header.class_size = classes_size / sizeof (struct dex_class_t);
 	bin->classes = (struct dex_class_t *) malloc (classes_size);
-	r_buf_read_at (bin->b, bin->header.class_offset, (ut8*)bin->classes, classes_size);
+	for (i = 0; i < bin->header.class_size; i++) {
+		bin->classes[i].class_id = r_read_le32 (bin->b->buf + bin->header.class_offset +
+						i * sizeof (struct dex_class_t) + 0);
+		bin->classes[i].access_flags = r_read_le32 (bin->b->buf + bin->header.class_offset +
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: what about use r_offsetof. and maybe deifne a macro to make all this stuff much more readable?

File: libr/bin/format/dex/dex.c [hunk 4]
```diff
@@ -107,23 +161,18 @@ struct r_bin_dex_obj_t* r_bin_dex_new_buf(RBuffer *buf) {
...
-			if (cur > 0x7f) {
-				cur = *(ptr++);
-				result |= (cur & 0x7f) << 21;
-				if (cur > 0x7f) {
-					cur = *(ptr++);
-					result |= cur << 28;
-				}
-			}
-		}
+	while(shift <= 28) {
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: missing space before (

Comment (radare) [inline libr/bin/format/dex/dex.c]: why not < 29 ? :D

File: libr/bin/format/dex/dex.c [hunk 5]
```diff
@@ -107,23 +161,18 @@ struct r_bin_dex_obj_t* r_bin_dex_new_buf(RBuffer *buf) {
...
+		byte = *(in--);
+		result |= (byte & 0x7f << shift);
+		if (byte > 0x7f)
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: missing braces

File: libr/bin/format/dex/dex.c [hunk 6]
```diff
@@ -140,27 +189,30 @@ int dex_uleb128_len (const ut8 *ptr) {
 
 #define SIG_EXTEND(X,Y) X = (X << Y) >> Y
 int dex_read_sleb128 (const char *ptr) {
-	int cur, result = *(ptr++);
+	int cur, result;
+	ut8 len = dex_uleb128_len(ptr);
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: missing space before (

File: libr/bin/p/bin_dex.c [hunk 7]
```diff
@@ -1047,17 +1066,10 @@ static RList* sections(RBinFile *arch) {
...
-	ret = r_buf_fread_at (arch->buf, 100, (ut8*)&off, "i", 1);
-	if (ret != 4) {
-		return 0;
-	}
-	ret = r_buf_fread_at (arch->buf, 104, (ut8*)&len, "i", 1);
-	if (ret != 4) {
-		return 0;
-	}
-	return off + len + 0x20;
+	off = r_read_le32 (arch->buf->buf + 108);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: This can oob. Use rbuf api

File: libr/bin/p/bin_dex.c [hunk 8]
```diff
@@ -818,9 +831,15 @@ static RList* classes (RBinFile *arch) {
 	for (i = 0; i < bin->header.class_size; i++) {
 		ut64 class_addr = (ut64) bin->header.class_offset \
 			+ (sizeof (struct dex_class_t)*i);
-		// ETOOSLOW
-		r_buf_read_at (bin->b, class_addr, (ut8*)&entry,
-			sizeof (struct dex_class_t));
+		entry.class_id = r_read_le32 (bin->b->buf + class_addr + 0);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Put those assignements in a separate function and maybe i would keep the old method in a #if block

File: libr/bin/p/bin_dex.c [hunk 9]
```diff
@@ -316,24 +312,28 @@ static char *getClassName(const char *name) {
...
 	DexField field = { 0 };
+	ut8 ff[sizeof (DexField)] = {0};
 	ut32 off = dex->header.fields_offset;
 	int delta = fn * sizeof (DexField);
 	if (delta > dex->header.fields_size) {
 		return NULL;
 	}
-	r_buf_read_at (dex->b, off, (ut8*)&field + delta, sizeof (DexField));
+	r_buf_read_at (dex->b, off + delta, ff, sizeof (DexField));
+	field.name_id = r_read_le32 (&ff[4]);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Use ff + 4

File: libr/bin/p/bin_dex.c [hunk 10]
```diff
@@ -316,24 +312,28 @@ static char *getClassName(const char *name) {
...
 	DexField field = { 0 };
+	ut8 ff[sizeof (DexField)] = {0};
 	ut32 off = dex->header.fields_offset;
 	int delta = fn * sizeof (DexField);
 	if (delta > dex->header.fields_size) {
 		return NULL;
 	}
-	r_buf_read_at (dex->b, off, (ut8*)&field + delta, sizeof (DexField));
+	r_buf_read_at (dex->b, off + delta, ff, sizeof (DexField));
+	field.type_id = r_read_le16 (&ff[2]);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Use f + 2

File: libr/bin/p/bin_dex.c [hunk 11]
```diff
@@ -452,13 +455,18 @@ static int *parse_class(RBinFile *binfile, struct r_bin_dex_obj_t *bin, struct d
 			free (methods);
 			return NULL;
 		}
+		r_buf_read_at (binfile->buf, bin->header.fields_offset +
+				fieldIdx * sizeof (DexField), ff, sizeof (DexField));
+		fields[fieldIdx].class_id = r_read_le16 (&ff[0]);
+		fields[fieldIdx].type_id = r_read_le16 (&ff[2]);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Ff+x

File: libr/bin/p/bin_dex.c [hunk 12]
```diff
@@ -818,9 +831,15 @@ static RList* classes (RBinFile *arch) {
 	for (i = 0; i < bin->header.class_size; i++) {
 		ut64 class_addr = (ut64) bin->header.class_offset \
 			+ (sizeof (struct dex_class_t)*i);
-		// ETOOSLOW
-		r_buf_read_at (bin->b, class_addr, (ut8*)&entry,
-			sizeof (struct dex_class_t));
+		entry.class_id = r_read_le32 (bin->b->buf + class_addr + 0);
+		entry.access_flags = r_read_le32 (bin->b->buf + class_addr + 4);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Use rbuf api to read the bytes you need to parae. Is not correct to directly point to stuff inside the suposed buffer. Which can be even null in some situations

File: libr/bin/p/bin_dex.c [hunk 13]
```diff
@@ -475,16 +483,21 @@ static int *parse_class(RBinFile *binfile, struct r_bin_dex_obj_t *bin, struct d
 		p = r_uleb128 (p, p_end - p, &fieldIdx); // fieldIdx
 		p = r_uleb128 (p, p_end - p, &accessFlags); // accessFlags
 		fieldIdx += lastidx;
+		r_buf_read_at (binfile->buf, bin->header.fields_offset +
+				fieldIdx * sizeof (DexField), ff, sizeof (DexField));
+		fields[fieldIdx].class_id = r_read_le16 (&ff[0]);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Use ff

File: libr/bin/p/bin_dex.c [hunk 14]
```diff
@@ -475,16 +483,21 @@ static int *parse_class(RBinFile *binfile, struct r_bin_dex_obj_t *bin, struct d
...
+		fields[fieldIdx].type_id = r_read_le16 (&ff[2]);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Use ff+2

**Comment by alvarofe** [comment]
@zamaudio here you have a patch to fix the crash

```
diff --git a/libr/bin/format/dex/dex.c b/libr/bin/format/dex/dex.c
index b75d9e4..a10c807 100644
--- a/libr/bin/format/dex/dex.c
+++ b/libr/bin/format/dex/dex.c
@@ -154,25 +154,27 @@ fail:
 }
 
 // Move to r_util ??
-int dex_read_uleb128 (const ut8 *ptr) {
+int dex_read_uleb128(const ut8 *ptr) {
 	ut8 len = dex_uleb128_len (ptr);
-	ut8 *in = ptr + len - 1;
+	const ut8 *in = ptr + len - 1;
 	ut32 result = 0;
 	ut8 shift = 0;
 	ut8 byte;
 
-	while(shift < 29) {
+	while(shift < 29 && len > 0) {
 		byte = *(in--);
 		result |= (byte & 0x7f << shift);
-		if (byte > 0x7f)
+		if (byte > 0x7f) {
 			break;
+		}
 		shift += 7;
+		len--;
 	}
 	return result;
 }
 
 #define LEB_MAX_SIZE 6
-int dex_uleb128_len (const ut8 *ptr) {
+int dex_uleb128_len(const ut8 *ptr) {
 	int i = 1, result = *(ptr++);
 	while (result > 0x7f && i <= LEB_MAX_SIZE) {
 		result = *(ptr++);
@@ -182,7 +184,7 @@ int dex_uleb128_len (const ut8 *ptr) {
 }
 
 #define SIG_EXTEND(X,Y) X = (X << Y) >> Y
-int dex_read_sleb128 (const char *ptr) {
+int dex_read_sleb128(const char *ptr) {
 	int cur, result;
 	ut8 len = dex_uleb128_len (ptr);
 	ptr += len - 1;
diff --git a/libr/bin/p/bin_dex.c b/libr/bin/p/bin_dex.c
index 86a04da..45277a2 100644
--- a/libr/bin/p/bin_dex.c
+++ b/libr/bin/p/bin_dex.c
@@ -140,7 +140,7 @@ static RBinInfo *info(RBinFile *arch) {
 	return ret;
 }
 
-static RList* strings (RBinFile *arch) {
+static RList* strings(RBinFile *arch) {
 	struct r_bin_dex_obj_t *bin = NULL;
 	RBinString *ptr = NULL;
 	RList *ret = NULL;
@@ -148,23 +148,28 @@ static RList* strings (RBinFile *arch) {
 	ut8 buf[6];
 	ut64 off;
 
-	if (!arch || !arch->o)
+	if (!arch || !arch->o) {
 		return NULL;
+	}
 	bin = (struct r_bin_dex_obj_t *) arch->o->bin_obj;
-	if (!bin || !bin->strings)
+	if (!bin || !bin->strings) {
 		return NULL;
+	}
 	if (bin->header.strings_size > bin->size) {
 		bin->strings = NULL;
 		return NULL;
 	}
-	if (!(ret = r_list_new ()))
+	if (!(ret = r_list_new ())) {
 		return NULL;
+	}
 	ret->free = free;
 	for (i = 0; i < bin->header.strings_size; i++) {
-		if (!(ptr = R_NEW0 (RBinString)))
+		if (!(ptr = R_NEW0 (RBinString))) {
 			break;
-		if (bin->strings[i] > bin->size || bin->strings[i] + 6 > bin->size)
+		}
+		if (bin->strings[i] > bin->size || bin->strings[i] + 6 > bin->size) {
 			goto out_error;
+		}
 		r_buf_read_at (bin->b, bin->strings[i], (ut8*)&buf, 6);
 		len = dex_read_uleb128 (buf);
 		if (len > 1 && len < R_BIN_SIZEOF_STRINGS) {
@@ -382,7 +387,7 @@ static char *dex_class_super_name (RBinDexObj *bin, RBinDexClass *c) {
 }
 
 static int *parse_class(RBinFile *binfile, struct r_bin_dex_obj_t *bin, struct dex_class_t *c, RBinClass *cls) {
-	int i, *methods;
+	int total, i, *methods;
 	ut64 SF, IF, DM, VM;
 	const ut8 *p, *p_end;
 	char *class_name;
@@ -451,12 +456,12 @@ static int *parse_class(RBinFile *binfile, struct r_bin_dex_obj_t *bin, struct d
 		p = r_uleb128 (p, p_end - p, &fieldIdx); // fieldIdx
 		p = r_uleb128 (p, p_end - p, &accessFlags); // accessFlags
 		fieldIdx += lastidx;
-		if (bin->header.fields_offset + sizeof (DexField) * fieldIdx > bin->size) {
+		total = bin->header.fields_offset + fieldIdx * sizeof (DexField);
+		if (total > bin->size || fieldIdx > SF) {
 			free (methods);
 			return NULL;
 		}
-		r_buf_read_at (binfile->buf, bin->header.fields_offset +
-				fieldIdx * sizeof (DexField), ff, sizeof (DexField));
+		r_buf_read_at (binfile->buf, bin->header.fields_offset + fieldIdx * sizeof (DexField), ff, sizeof (DexField));
 		fields[fieldIdx].class_id = r_read_le16 (ff);
 		fields[fieldIdx].type_id = r_read_le16 (ff + 2);
 		fields[fieldIdx].name_id = r_read_le32 (ff + 4);

```

**Comment by radare** [comment]
Thanks! there's just a missing space before a (, but it's ok :) we will be fixing syntax issues forever :P

**Pull request merged**